### DISCUSSION TITLE
Discover and auto pair with password

### DISCUSSION THREAD

**Discussion by darkbox** [bluez/bluer#70] [General]
Hi,
I don't know if this is possible but I made a small program to connect to a solar inverter and right now I'm using `bluetoothctl` with the help of a script to auto discover the device and pair it. ¿Is there any way to do this from directly from rust code? So I can avoid using cron, hacky scripts, etc just to pair a device?

Thank you in advance.

**Comment by surban** [comment]
Yes. See the agent code in `gattcat` for example.

**Reply by darkbox** [reply]
I must been doing something wrong. I get "Error: Bluetooth authentication failed: Authentication Failed" when `device.pair().await?;` is called.
I've registered a custom agent as follows:
```
let agent = Agent {
    request_pin_code: Some(Box::new(|req| Box::pin(request_pin_code(req)))),
    request_passkey: Some(Box::new(|req| Box::pin(request_passkey(req)))),
    ..Default::default()
};
let _ = session.register_agent(agent);
```
I used both pin code and passkey because I'm not sure which of them should be used.
```
async fn request_pin_code(_req: RequestPinCode) -> ReqResult<String> {
    Ok("123456".into())
}
async fn request_passkey(_req: RequestPasskey) -> ReqResult<u32> {
    Ok(123456)
}
```

**Reply by surban** [reply]
`let _ = ...` leads to immediate drop of the provided value, thereby unregistering the agent.

You must keep the returned handle alive for the agent to work.

**Reply by darkbox** [reply]
Thanks a lot! that was the issue.

**Comment by darkbox** [comment]
I can not make it work. I need some help 🙏.
To expand a little bit, `bluetoothctl` gives me almost the same error only if I try to pair it with the next command I get: 
```bash
$ bluetoothctl pair 48:70:1E:53:38:FC
Attempting to pair with 48:70:1E:53:38:FC
Failed to pair: org.bluez.Error.AuthenticationFailed
```
But when doing it manually like so it works perfectly. **How can I replicate this behavior using Bluer?**
```bash
$ bluetoothctl
Agent registered
[CHG] Controller B0:10:41:23:37:A6 Pairable: yes
[ID92932106100853      ]# pair 48:70:1E:53:38:FC
Attempting to pair with 48:70:1E:53:38:FC
Request passkey
[agent] Enter passkey (number in 0-999999): 123456
[CHG] Device 48:70:1E:53:38:FC Paired: yes
Pairing successful
[ID92932106100853      ]#
```
